Add models, controllers and tests

Vanchey359 opened the pull request:
Добавил модели, контроллеры к ним, написал тесты. В постмане все проходит.
Первый пул реквест завалил проверку гита из за 17той версии jdk, сейчас поменял в pom.xml все под 11, надеюсь пройдет :)

Vanchey359 committed:
add spring boot project

Vanchey359 committed:
add models, controllers and tests

Vanchey359 committed:
add models, controllers and tests (and change java 17 to 11)

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,54 @@
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+        </dependency>
```
Форматируй файл пробелами вместо табов, это можно сделать в несколько кликов в intellij

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -0,0 +1,13 @@
+@SpringBootApplication
+public class FilmorateApplication {
+
+	public static void main(String[] args) {
```
Форматируй пробелами

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+
+    @GetMapping
+    public List<Film> findAll() {
+        log.info("Получен список всех фильмов");
```
Логирование только на английском

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+@RequestMapping("/films")
+public class FilmController {
+
+    private final List<Film> films = new ArrayList<>();
```
Используй хэшмап для хранения фильмов, как мы делали в прошлых спринтах

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+    public Film updateFilm(@Valid @RequestBody Film film) {
+        validate(film);
+        for (Film fl : films) {
+            if (fl.getId() == film.getId()) {
```
Нужно будет использовать `Map#contains` чтобы выяснить, есть ли уже такой фильм

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void validate(Film film) {
+        if (film.getName() == null || film.getName().isBlank()) {
```
Это валидация уже делается посредством аннотации `@NotBlank`, здесь вручную проверять нет необходимости

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,54 @@
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-parent</artifactId>
+		<version>2.7.8</version>
+		<relativePath/> <!-- lookup parent from repository -->
```
Удали коммент

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,54 @@
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+		</dependency>
+
```
Зачем здесь пустая линия?

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,54 @@
+			<artifactId>spring-boot-starter-validation</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>jakarta.validation</groupId>
```
Выше уже есть эта же зависимость, эту декларацию удали, в импортах пакет `jakarta` замени на `javax`. Не забудь обновить мавен в intellij

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+            throw new ValidationException("Film title cannot be empty");
+        }
+
+        if (film.getDescription().length() > 200) {
```
Тоже самое, проверяется аннотацией `@Size`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        if (film.getDescription().length() > 200) {
+            log.warn("Запрос на действие с фильмом длинна описания которого больше 200 символов");       // Не особо понял зачем я это пишу, ведь когда программа выбросит исключение, там будет сообщение с причиной ошибки
```
Да, предлагаю здесь не логировать, логировать следует когда перехватываешь исключение, а не когда бросаешь

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+            throw new ValidationException("Release date must not be earlier than 28.12.1895");
+        }
+
+        if (film.getDuration() <= 0) {
```
Уже проверяется аннотацией `@Positive`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,77 @@
+            throw new ValidationException("Date of birth cannot be in the future");
+        }
+    }
+}
```
Те же замечания, что и для фильм-контроллера. Плюс не вижу логики про имя юзера и логин

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+        return film;
+    }
+
+    public void validate(Film film) {
```
private, тестировать через публичные методы

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,22 @@
+@Data
+public class Film {
+    private int id;
+    @NotBlank
```
Так как здесь аннотации, лучше отделить поля пустыми линиями, легче читается когда много полей

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,22 @@
+    private int id;
+    @NotBlank
+    private String name;
+    @Size(min = 1, max = 200)
```
Добавь `@NotNull`, `@Size` на null не проверяет

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,21 @@
+    @NotBlank
+    private String login;
+    private String name;
+    @PastOrPresent
```
Добавь `@NotNull`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,21 @@
+public class User {
+    private int id;
+    @NotBlank @Email
+    private String email;
```
```
@NotBlank
@Email
private String email;
```

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,109 @@
+	void shouldValidateFilmWithBigDescription() {
+		Film film = new Film();
+		film.setName("Test");
+		film.setDescription("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff" +
```
`"fff".repeat(100);`

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,109 @@
+@SpringBootTest
+class FilmorateApplicationTests {
+
+	static FilmController filmController = new FilmController();
```
Форматируй пробелами

Vanchey359 committed:
Bug fixes

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+                films.put(film.getId(), film);
+                log.info("Film {} has been updated", film.getName());
+            } else {
+                log.warn("Film with this id was not found");
```
Предлагаю здесь не логировать

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        if (film.getDuration() <= 0) {
+            throw new ValidationException("Film duration must be positive");  // Сначала удалил такие места где аннотации в model перекрывают эту валидацию, при написании тестов увидел что все таки не перекрывают
```
При обычных тестах java bean валидации не будут работать, но будут работать по обращении через http

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+
+        if (film.getDuration() <= 0) {
+            throw new ValidationException("Film duration must be positive");  // Сначала удалил такие места где аннотации в model перекрывают эту валидацию, при написании тестов увидел что все таки не перекрывают
+            // например тут в тесте я вводил dutation -4 и тест писал что "ожидается исключение ... но никакого исключения нет. Вернул эту валидацию обратно, тест работает
```
Удали валидацию, дублирующую валидации аннотациями и удали тесты, которые проверяют эти кейсы. Если не ошибаюсь, здесь должна остаться только валидация даты релиза

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,69 @@
+            throw new ValidationException("Login cannot be empty and contain spaces");
+        }
+
+        if (user.getName() == null || user.getName().isBlank()) {
```
Этот блок не является валидацией, перемести его отсюда

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,111 @@
+    @Test
+    void shouldValidateFilmWithEmptyNameAndEmptyAllFields() {
+        Film film = new Film();
+        Exception exception = assertThrows(NullPointerException.class, () -> filmController.addFilm(film));
```
Никогда не надо тестировать NPE, если он вылетел, это всегда ошибка. Выясни откуда он у тебя там вылетает и поправь

Тестировать необходимо только те исключения, которые мы сами выбрасываем

Vanchey359 replied:
NPE вылетает из за того что поля помеченные NotNull не заполнены, странно но эта аннотация почему то в тестах срабатывает "_"

Vanchey359 committed:
Fixed validation and testing.
I test validation with my own method with unit tests, validation with annotations in postman - all tests passed.

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+    @PutMapping
+    public Film updateFilm(@Valid @RequestBody Film film) {
+        validate(film);
+        if (films.containsKey(film.getId())) {
```
Guard clause

```
if (!films.containsKey(film.getId())) {
    throw ...
}
films.put(...
```

https://www.codementor.io/@clintwinter/use-guard-clauses-for-cleaner-code-1rrsczgwxp

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,67 @@
+
+    @PutMapping
+    public User updateUser(@Valid @RequestBody User user) {
+        if (users.containsKey(user.getId())) {
```
1. Нет вызова `validate`
2. guard clause

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,52 @@
+@SpringBootTest
+class FilmorateApplicationTests {
+
+    static FilmController filmController = new FilmController();
```
А почему static?

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,52 @@
+    static UserController userController = new UserController();
+
+    @Test
+    void shouldValidateFilmWithOldReleaseDate() {
```
Сначала должно идти название тестируемого метода

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Test
+    void shouldValidateUserWithEmptyName() {
```
`addUserShouldSetLoginToNameIfNameIsEmpty`

Vanchey359 committed:
bug fixes
